codex/sales/states/filled.nim

```diff
@@ -0,0 +1,46 @@
+import pkg/questionable
+import ./errored
+import ./finished
+import ./cancelled
+import ./failed
+import ../statemachine
+
+type
+  SaleFilled* = ref object of SaleState
+  SaleFilledError* = object of CatchableError
+
+method onCancelled*(state: SaleFilled, request: StorageRequest) {.async.} =
+  await state.switchAsync(SaleCancelled())
+
+method onFailed*(state: SaleFilled, request: StorageRequest) {.async.} =
+  await state.switchAsync(SaleFailed())
+
+method `$`*(state: SaleFilled): string = "SaleFilled"
+
+method enterAsync(state: SaleFilled) {.async.} =
+  without agent =? (state.context as SalesAgent):
+    raiseAssert "invalid state"
+
+  try:
+    let market = agent.sales.market
+
+    without slotIndex =? agent.slotIndex:
+      raiseAssert "no slot selected"
+
+    if availability =? agent.availability:
+      agent.sales.remove(availability)
```

Comment:
Availability should be removed much earlier; no two agents should start with the same availability otherwise you might end up selling your available store more than once :)

Reply:
Moved this to the downloading state. This should work across node restarts once the availability state is persisted on disk.